perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_userCertificates.java

```diff
@@ -0,0 +1,103 @@
+package cz.metacentrum.perun.core.impl.modules.attributes;
+
+import cz.metacentrum.perun.core.api.Attribute;
+import cz.metacentrum.perun.core.api.AttributeDefinition;
+import cz.metacentrum.perun.core.api.AttributesManager;
+import cz.metacentrum.perun.core.api.User;
+import cz.metacentrum.perun.core.api.exceptions.AttributeNotExistsException;
+import cz.metacentrum.perun.core.api.exceptions.ConsistencyErrorException;
+import cz.metacentrum.perun.core.api.exceptions.InternalErrorException;
+import cz.metacentrum.perun.core.api.exceptions.WrongAttributeAssignmentException;
+import cz.metacentrum.perun.core.impl.PerunSessionImpl;
+import cz.metacentrum.perun.core.implApi.modules.attributes.UserVirtualAttributesModuleAbstract;
+import cz.metacentrum.perun.core.implApi.modules.attributes.UserVirtualAttributesModuleImplApi;
+
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This attribute module is used to filter user certificates only to ten newest according to certificates expiration
+ */
+public class urn_perun_user_attribute_def_virt_userCertificates extends UserVirtualAttributesModuleAbstract implements UserVirtualAttributesModuleImplApi {
+
+
+	@Override
+	public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDefinition attributeDefinition) {
+		Attribute resultAttribute = new Attribute(attributeDefinition);
+
+		Attribute userCertsAttribute = getUserCertsAttribute(sess, user);
```

Comment:
I know that this is clear and correct implementation, but we should minimize number of selects. Like retrieve both attributes at once. Also dependency on another virtual attributes is problematic. As a result, you twice retrieve value of ``user:def:userCertificates`` and then do some work with it. Maybe you should extract logic from ``getValue()`` of ``user:def:userCertificates`` and make it ``ModulesUtilsBlImpl`` method. Then you could use it directly in both modules to select user attributes only once and construct expiration mapping locally in this module (without dependency on virtual attribute).

Reply:
Reworked